### PR TITLE
Migrate from Truth to AssertJ

### PR DESCRIPTION
**Pull request by anuraaga**
See #1452 for motivation, this should be a benefit for new contributor experience.

I used IntelliJ replace in path to replace imports and some obvious differences like `.isSameInstanceAs` -> `.isSameAs` and manually updated map assertions for `entry` (assertj is typesafe) and fixed failing list assertions to assert with any order, truth defaults to allowing any order and assertj defaults to exact order.

### CODE REVIEW

**Comment by carlosalberto** [comment]
Hey hey -is there any technical advantage for this? I see a few comments from the referenced issue but it sounds like popularity is the most important factor here?

> AssertJ is extremely popular in the Java ecosystem while Truth is mostly used by Google projects. 

Well, OC (and a lot of components we use) were created by Google, and I see no problem with that, unless they are inferior in quality or have been deprecated). To me, migrating to JUnit 5 is a great thing, but this change feels like a matter of personal taste (not to mention AssertJ and Truth are very similar API-wise, and also that, I think, new contributors won't care much about whether we using one or the another).

For example: I went and looked for a comparison and found issues like [this issue](https://github.com/checkstyle/checkstyle/issues/7239) or [this article](https://truth.dev/comparison.html), for example, to have a counter point for migrating to AssertJ.

**Comment by anuraaga** [comment]
There's no technical reason. Which is why popularity should be a major factor here. Because the number of forks gets counted it's not a prcise measurements but seeing _two orders of magnitude_ difference in dependants is a trend I think. We can expect this because in Java Spring is huge, and Spring has adopted AssertJ. Without a significant technical reason to pick something else, we should lean towards what has a higher probability for contributors to be familiar with.

> new contributors won't care much about whether we using one or the another

I cared because I've used assertj when contributing to several other projects and Truth, only at Google - remembering the differences in API was extra effort. I can't speak for everyone but the idea is the trend indicates there is high probability of this happening more widely. I would not make such a change out of personal preference only, my opinion is that using Truth is personal preference and missing the larger trends of the community.

In #529, the only reason given to stick to truth was the prevalence in this codebase. But now you have a PR that migrates it all so that's not a reason. Please take such efforts into account

**Comment by kenfinnigan** [comment]
From a long time open-source contributor to numerous Java projects, I can attest that AssertJ is widely used among Java frameworks and applications.

Until I saw this thread, I'd never heard of Truth before.

**Comment by Oberon00** [comment]
What I like about AssertJ is that you can use `as` and `withFailureMessage` to get nicer exception messages. But I don't think that's a strong argument.

**Comment by carlosalberto** [comment]
> There's no technical reason. Which is why popularity should be a major factor here.

I'm on the boat of "if it's not broken, don't fix it". On the other hand, given the fact also that Google has and is putting resources into OTel (like the recent zPages support), I see a good thing their developers can come and simply use what they are used to, as a counter-example.

Testing is not affecting final users, and we have only so many contributors at this moment. Given that @jkwatson has being vocal on the lack of cycles we have, I'd personally have put effort into other items myself ;)

I'm not against this PR but I'm definitely not a fan of it, as to me, it's a matter of personal taste (and I'm afraid of a Google contributor coming in the future to switch back to Truth, because of his own personal taste as well). Please go ahead and merge when/if other approvers are fine with it though ;)

**Comment by jkwatson** [comment]
I'd rather make the repo more friendly to the community at large, rather than cater to a single vendor's preference, so I'm in favor of switching to AssertJ because it is better known to the majority of the community.

**Comment by bogdandrutu** [comment]
@jkwatson 

>I'd rather make the repo more friendly to the community at large, rather than cater to a single vendor's preference, so I'm in favor of switching to AssertJ because it is better known to the majority of the community.

I don't think we are talking about any vendor here, it is just a library preference, as you can see the Truth was a huge influence to the `AssertJ` which changed their style to be like "assertThat.().isFoo()". So we are talking about using the library that makes the code more readable, as you can see I approved this because I feel with the latest changes `AssertJ` made huge improvements in readability

File: api/src/test/java/io/opentelemetry/trace/TraceStateTest.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ void getEntries() {
     assertThat(secondTraceState.getEntries())
         .containsExactly(TraceState.Entry.create(SECOND_KEY, SECOND_VALUE));
     assertThat(multiValueTraceState.getEntries())
-        .containsExactly(
+        .containsExactlyInAnyOrder(
```
Comment (jkwatson) [inline api/src/test/java/io/opentelemetry/trace/TraceStateTest.java:55]: isn't the order actually important here?

File: sdk/build.gradle [hunk 2]
```diff
@@ -18,6 +18,7 @@ dependencies {
     annotationProcessor libraries.auto_value
 
     testAnnotationProcessor libraries.auto_value
+    testCompileOnly libraries.auto_value_annotation
```
Comment (jkwatson) [inline sdk/build.gradle:21]: any idea why this was needed to be added? seems unrelated to the PR, on its face.

File: api/src/test/java/io/opentelemetry/trace/TraceStateTest.java (hunk 1 above)
Comment (anuraaga) [reply api/src/test/java/io/opentelemetry/trace/TraceStateTest.java:55]: Confirmed that the order seems to be changing here - filed #1499 will look into it in a follow up.

File: sdk/build.gradle (hunk 2 above)
Comment (anuraaga) [reply sdk/build.gradle:21]: I think Truth was leaking this in their POM so the test dependency on Truth got this transitively.

File: api/src/test/java/io/opentelemetry/trace/TraceStateTest.java (hunk 1 above)
Comment (jkwatson) [reply api/src/test/java/io/opentelemetry/trace/TraceStateTest.java:55]: I think the order is supposed to be the opposite of what you put in, if you look at the javadoc on the TraceState.Builder class. I think this test just wasn't asserting that fact.

Comment (anuraaga) [reply api/src/test/java/io/opentelemetry/trace/TraceStateTest.java]: Ah you did the digging for me! Thanks I went ahead and updated the test here.

Comment (jkwatson) [reply api/src/test/java/io/opentelemetry/trace/TraceStateTest.java]: I do like that assertJ makes it super clear when you aren't enforcing the order, though!
